### DISCUSSION TITLE
Support arbitrary repository types for private cocoapods datasource

### DISCUSSION THREAD

**Discussion by mitchins** [renovatebot/renovate#29049] [Suggest an Idea]
### Problem Definition
Our organisation uses Mend/Renovate however not being able to query our internal packages for automatic PR can be limiting - particularly because we mirror third party packages internally for security purposes as we are a bank.

From what I can tell, the Cocoapods datasource should in theory work with a private repository - assuming it's:
(1) On Github
(2) Structured in the way that is expected (with the content and specs folders)

It does not however work with any other repository type as it uses the github v3 api to query contents and packages. After looking through the source and building locally - I think it can be handled fairly elegantly, but would hope that the work could be accepted upstream if I goto the effort of supporting it.

### Rationale for the suggested enhancement
(A) The organisation I work for uses Azure Devops (We won't be moving to GitHub actions anytime soon either) however manyorganisations still use BitBucket or Gitlab.
(B) Cocoapods is still a popular package manager even with Swift Package Manager as it's much more powerful and proven for enterprise and advanced usages.

### Suggested approach
Assumptions:
(1) There are too many different git providers besides GitHub to handle their HTTP APIs
(2) We would have permissions to check out the repository holding the podspecs, given we are supposed to update packages
(3) A shallow checkout of depth=1 should be quick enough, and we can clean it up once done

With the above, if the repository is not GitHub, e.g. dev.azure.com or myorg.bitbucket.com (or bitbucket.myorg.com), rather than what it does now (which is to try to call the GitHub api and fail with 404/401), let's just check out the repository locally using a shallow clone, and query the directory structure. 

This approach is more or less how cocoapods itself queries it, and even caches it locally (although we should let it do its thing rather than try to inspect that).

### In Summary
For the index.ts of the Cocoapods DataSource, if the repository url is not the public CDN or a github repository, can we use a strategy of checking out the repository to query available packages in the repository structure? This pattern isn't new either as it looks like Cargo (among others) uses a repository clone. 

The approach would work seamlessly with any "other" provider besides the CDN or GitHub, and as mentioned there's still alot of life left in CocoaPods as well as enterprise usage of non GitHub repositories.

Again while I am willing to look into the work, I'd prefer to have an approach that could be accepted upstream so that it could:
(A) Work with Mend cloud service
(B) Not require manual merges from upstream

**Comment by rarkins** [comment]
Do you mean that you mirror https://github.com/CocoaPods/Specs? And that you'd want to configure Renovate to tell it that using `packageRules` with `registryUrls`?

Regarding:
> (1) There are too many different git providers besides GitHub to handle their HTTP APIs

We have a getJson function for all our supported platforms, including Azure DevOps. It's currently used for fetching presets. So I don't think this assumption holds.

**Reply by mitchins** [reply]
Thanks for your reply.

We don't mirror cocoa pods, we:

Have our own private packages for internal libraries in the format:
`/<Library>/<Version>/<Library>.podspec.json`
That is the standard format for a private repository as sharding etc isn't require - so it's the simple way.

Most of these libraries are internal modules but we have some vendor packages hosted too.

At the top of our Podfile we declare the repository:

`Source "https://dev.azure.com/myOrg/myProject/_git/pod-specs-repository/`

With regards to Azure, it supports fetching the actual projects from azure devops just fine. However cocoapods data source can only fetch from GitHub.
The Datasource assumes it is either official CDN or GitHub, and it uses the GitHub HTTP api only to access it. This isn't the same API as azure.

I looked in the http component folder and I didn't see a class that exposes the same function for azure as GitHub with getJson.

From here:

https://github.com/renovatebot/renovate/blob/04f4edbd518f4fe7409a573357efed3fa8ff82a9/lib/modules/datasource/pod/index.ts

It has a reference to:

`githubHttp: GithubHttp;` in the code below it makes use of `.getJson`

However, I cannot see an equivalent class or function for azure based on the above.

Called here:

```
function releasesGithubUrl(
  packageName: string,
  opts: {
    hostURL: string;
    account: string;
    repo: string;
    useShard: boolean;
    useSpecs: boolean;
  },
): string {
  const { hostURL, account, repo, useShard, useSpecs } = opts;
  const prefix =
    hostURL && hostURL !== 'https://github.com'
      ? `${hostURL}/api/v3/repos`
      : 'https://api.github.com/repos';
  const shard = shardParts(packageName).join('/');
  // `Specs` in the pods repo URL is a new requirement for legacy support also allow pod repo URL without `Specs`
  const packageNamePath = useSpecs ? `Specs/${packageName}` : packageName;
  const shardPath = useSpecs
    ? `Specs/${shard}/${packageName}`
    : `${shard}/${packageName}`;
  const suffix = useShard ? shardPath : packageNamePath;
  return `${prefix}/${account}/${repo}/contents/${suffix}`;
}

private async getReleasesFromGithub(
    packageName: string,
    opts: { hostURL: string; account: string; repo: string },
    useShard = true,
    useSpecs = true,
    urlFormatOptions: URLFormatOptions = 'withShardWithSpec',
  ): Promise<ReleaseResult | null> {
    const url = releasesGithubUrl(packageName, { ...opts, useShard, useSpecs });
    const resp = await this.requestGithub<{ name: string }[]>(url, packageName);
    if (resp) {
      const releases = resp.map(({ name }) => ({ version: name }));
      return { releases };
    }

    // support different url formats
    switch (urlFormatOptions) {
      case 'withShardWithSpec':
        return this.getReleasesFromGithub(
          packageName,
          opts,
          true,
          false,
          'withShardWithoutSpec',
        );
      case 'withShardWithoutSpec':
        return this.getReleasesFromGithub(
          packageName,
          opts,
          false,
          true,
          'withSpecsWithoutShard',
        );
      case 'withSpecsWithoutShard':
        return this.getReleasesFromGithub(
          packageName,
          opts,
          false,
          false,
          'withoutSpecsWithoutShard',
        );
      case 'withoutSpecsWithoutShard':
      default:
        return null;
    }
  }

```

What would be a great help is 
(A) point me towards the equivalent azure function if I've missed it.
(B) indicate the preferred method of refactoring the existing cocoapods datasource to support the repository.

Cocoapods itself can use any GIT repository over HTTPS or SSH.
However it appears that because the repository is being queried with a platform specific api and not cloned, the current implementation for a data source won't work on anything besides GitHub.

Any help or direction would be much appreciated, however after extensive testing I've been unable to have renovate work with an azure repository for the cocoapods spec (even though pod itself works fine).

**Reply by rarkins** [reply]
The file fetching is in lib/modules/platform, such as https://github.com/renovatebot/renovate/blob/23421c572cd5aaa98bf0237250c305e51809d736/lib/modules/platform/azure/index.ts#L123-L180

This can be used if you're fetching from the same AZDO org as the repos Renovate is running against - is that the case?

An example of it being used in the presets logic is in https://github.com/renovatebot/renovate/blob/23421c572cd5aaa98bf0237250c305e51809d736/lib/config/presets/local/common.ts#L7-L33

The logic could (hopefully) be like this:

- If it's not the CDN and not GitHub
- Is it the same server as "endpoint"?
- If so, call platform.getRawFile() with appropriate params

**Reply by mitchins** [reply]
Thanks, yes it is the same platform/organisation azure/org (although the repository is different).
I will dig into what platform is, and I agree with your logic, if it’s not known as CDN or GitHub then fall back to pulling platform.

Although we would need to use “get file listing” as the versions are in hierarchical directory format, is this part of platform?

**Reply by rarkins** [reply]
File listing is not an exposed API like getRawFile(), but for which step do you need it?

BTW if you think it's necessary to use git, and the approach taken in Cargo would work, you are welcome to implement that.

**Comment by mitchins** [comment]
Great thanks.GIT would be the perfect “catch all”.I doubt performance of a couple extra seconds would matter, with shallow commit.Just wanted to get a sense check before I commit significant effort. With regards to GitHub currently it doesn’t actually check the domain, I believe “git” just needs to be in the host url regex. Obviously only GitHub supports the GitHub query API.I’m unsure what forms the domain takes with enterprise accounts, eg, does GitHub always occur even for big enterprises? Long story short:If url == CDN {}else if url == GitHub {}else { do GIT checkout and collect versions}As long as that seems endorse im happy to go ahead - just wanted to sense check.Given that cocoapods requires directory hierarchy to list versions this is not standardised, so apart from GitHub unless you want to support every sundry provider then a shallow clone is guaranteed to work (although not super elegant).Sent from my iPhoneOn 14 May 2024, at 19:28, Rhys Arkins ***@***.***> wrote:﻿
File listing is not an exposed API like getRawFile(), but for which step do you need it?
BTW if you think it's necessary to use git, and the approach taken in Cargo would work, you are welcome to implement that.

**Comment by mitchins** [comment]
@rarkins I have a PR ready, is it possible to get this turned into an issue so I can link towards it, or is there another way to go about it?
It appears that I'm not able to just open issues directly.

**Reply by rarkins** [reply]
Please write up your proposed solution so that it can be converted to an issue. Right now it's too "rambling" to copy all of this content into an issue - we need a summary/action.

Also I saw in your PR that you seem to have introduced a proprietary env variable for solving this, which is definitely not a way we want to solve it, so that's why we need to discuss it here a bit further.

**Reply by mitchins** [reply]
Understood.
### Here's a summary of the issue:

Although the pod datasource can detect the url of any arbitrary repository, attempting to read it assumes that is both on GitHub (uses GitHub v3 API) and that it is in the sharded format. In short, it can't read any private repository created as per the standard method on a basic git repo.

### Here's the solution proposal:

Currently the code assumes it's either Cocoapods CDN or Github hosting based on the URL, if github else CDN.
This isn't very flexible and any attempt to coerce private repositories based URL patterns will likely break existing functionality or not create the desired result.

The intended solution will (via configuration) determine that the registryURL is  private hosted repository and act accordingly by pulling down the repository and extracting versions. This is how cocoapods itself extracts versions when running pod and is universally compatible.

This does mean that we need a method to inject the predetermined private repository URLs, as configuration should be preferred over convention. As far as I am aware the cocoapods or `pod` datasource and manager don't have any configuration fields. If we can add one that's fine, we just need a way to specify that one (or more) custom repositories are used explicitly.

If that isn't possible than the last resort would be to check if the repository matches the github regex, then instead of assuming that it is the standard repository use an equality to check it, and if it isn't assume it's a private repository (given it would attempt to pull down the CDN and fail anyway), however this reduces the available range of private repository urls to those that don't match the current GitHub regex. 


Let me know if anything else is required, the actual solution isn't overly complex.

**Reply by mitchins** [reply]
To elaborate on the above, based on my experimentation in the `Podfile`:

```ruby
# The following will be treated by Renovate as "Github"
source 'https://qdigitalcode.visualstudio.com/BOQ%20Group%20Mobile%20Applications/_git/ios-specs-cocoapods-test'
# The following will be treated by Renovate as "CDN"
source 'https://qdigitalcode@dev.azure.com/qdigitalcode/BOQ%20Group%20Mobile%20Applications/_git/ios-specs-cocoapods-test'
```

The first one passes the regex for Github and it attempts to use GitHub v3 api, the second one falls through and it is assumed to be the CDN via https, which also fails as neither uses the expected protocol or format.

In my case, both are valid, and so I could feasibly use the second URL and so if it's not architecturally feasible to add a new configuration parameter to tell it what the private repository is, we could:

```
if registryURL =~ GitHubRegex 
# Do stuff for GitHub Releases
elif registryURL === publicCDN
# Do Stuff for cdn.cocoapods.org
else
# Assume its a private repo
endif
```

The ideal solution, and that which is in the PR has logic as:

```
if config.privateCocoapodsURLs.contains(registryURL)
# Do stuff for private repo
elif registryURL =~ GitHubRegex 
# Do stuff for GitHub Releases
else
# Assume its public CDN
endif
```

Which is simply adding an if before the existing if/else (converting the first if into elif).
The whole thing comes down to "Can we add a new parameter into the platform for specifying the exact custom repositories".
If the answer is no, then either we do the above, or we do something like add regex for bitbucket, gitlab, azure which should work for maybe 90% of people?

**Reply by mitchins** [reply]
In an effort to be and precise, short and not-rambling:

We need to determine when to pull from a private cocoapods registry via standard GIT as opposed to github or CDN.

I need endorsement for one of the following solutions:
* Create a new configuration variable to explicitly dictate one or more urls for private cocoapods registries. [Works in any case]
* Define well-known git providers who are not github, because github/cdn protocol WILL fail [Works in most cases]
* Assume when existing GitHub regex doesn't match, and != public CDN, we use private repo logic [Works in some/half cases]

Please keep in mind, it would not be breaking established precedence to define other possible git providers (in an array most likely) already explicitly has a regex for determining when it is specifically GitHub:
![image](https://github.com/renovatebot/renovate/assets/1016050/39a39315-f44e-49e5-a8a9-1a40c56a5851)
The advantage of defining a match for other GIT providers would allow it to check first and match gitlab.com and perform correct logic even if the existing GitHubregex would capture it today.

**Comment by ElliottJW** [comment]
Similarly, we are sourcing cocoapods from a private GitLab repo using specs, and Artifactory (not using specs file here) that Renovate resolves as GitHub and CDN URLs:

```
git@git.lab.domain.com:my-cocoapods/specs.git

# ...resolves as CDN:
git@git.lab.domain.com:my-cocoapods/specs.git/all_pods_versions_e_4_d.txt

# ====

https://artifactory.internal.domain.com/artifactory/api/pods/__our_local_cocoapods

# ...attempts to resolve as GitHub URL:
https://artifactory.internal.domain.com/api/v3/repos/artifactory/api/contents/Specs/e/a/4/my-cocoapod
```

Before finding this discussion, we were considering adding an `artifactory` URL regex (`/artifactory/api/pods/`) to determine if any of the registryUrls are based in [artifactory](https://jfrog.com/help/r/jfrog-artifactory-documentation/use-the-pod-command-line-cdn) for an MVP approach.

**Reply by mitchins** [reply]
@ElliottJW im curious how the artifactory setup works. Are spec files still resolved from GIT, but then the packages are pulled from artifactory?
We do the same thing but with azure universal packages (it requires a plugin)

**Reply by ElliottJW** [reply]
Speaking strictly for pods, Artifactory can be configured as a CDN or you can use the [cocoapods-art plugin](https://jfrog.com/help/r/jfrog-artifactory-documentation/use-the-cocoapods-art-plugin) to access the packages that are stored in Artifactory. Ours is the latter. The `__our_local_cocoapods` repo contains a `/.specs` file which contains the pods and their respective podspecs, similar to how GitHub has `Specs.git`

**Reply by mitchins** [reply]
So it sounds like we have the same problem. A basic GIT repository hosts the collection of specs files, is that right?

What does your structure look like?

We use 
<root>
  MyLibrary
    <Version>
      MyLibrary.podspec.json

So the list of versions is basically the glob status of version folders for that library containing a .podspec.json that matches.

That logic will need to be added into my PR, so if it's different I'll need to factor it in

**Reply by ElliottJW** [reply]
> A basic GIT repository hosts the collection of specs files

To be super specific: Are you referring to Artifactory when you say a git repository? I'd characterize Artifactory as more of an artifact storage by version rather than a version control system. Just want to make that distinction for anyone else reading this.

Our structure on Artifactory is:

`.../__our_local_cocoapods/.specs/<pod>/<version>/<pod>.podspec`

For example:

`.../__our_local_cocoapods/.specs/My-Cool-Lib-5.2/3.2.67/My-Cool-Lib-5.2.podspec`

>  the list of versions is basically the glob status of version folders for that library containing a .podspec.json that matches

Yes, each folder under the pod is a version, and each version folder holds the podspec.

**Reply by mitchins** [reply]
Thanks for helping tease this one out.

Let's back up a step (just because I'd like to ensure the solution works for everyone).

Here's our setup:
We store the spec files in GIT, and our project Podfile imports it by declaring:

`Source "https://dev.azure.com/myOrganisation/myTeam/cocoapods-specs/_git"`

Pod uses GIT to pull down the repository to sync the list of files, the index of this file could be:

LibraryA/1.0.0/LibraryA.podspec.json
LibraryA/1.0.1/LibraryA.podspec.json
LibraryB/1.0.0/LibraryB.podspec.json

_Side-note: Each podspec declares the `source` for the particular library, in our case that's a https location hosted on azure universal packages. (It could be any https source or a git source, but in our case we use azure packages + the official plugin to support it)._

I'd assumed that when you said artifactory it merely hosted the actual packages like our package channel on azure.

What's most relevant to the PR I am trying to raise is that we need to understand how the list of packages is fetched.
If cocoapods support it, then so should renovate.

So with that in mind, if you could elaborate - providing any examples of hierarchy or pod source declarations it would to a long way to ensuring that I create a PR that works for as many people as possible.
Even if our organisation doesn't use it, I could still install artifactory via docker to emulate it.

**Reply by ElliottJW** [reply]
> I'd assumed that when you said artifactory it merely hosted the actual packages like our package channel on azure.

👍 I'm nearing the limit of my understanding of iOS and Artifactory, coming from Android, so I'll try to explain to the best of my ability.

>  examples of ... pod source declarations

In our Podfile, we use Artifactory's plugin to source our spec + binaries:

```
# Podfile
source 'https://github.com/CocoaPods/Specs.git'
source 'git@git.lab.instance.com:my-cocoapods/specs.git'

# To get pods from Artifactory:
plugin 'cocoapods-art', :sources => [
    'artifactory-private'   # A private specs+binaries repository on Artifactory
]
```

>  examples of hierarchy

Artifactory manages the folder structure of the packages and applies the source in the podspec on our behalf. The packages are organized in a similar format though:

General example:
`.../__our_local_cocoapods/our-team/<pod>/<version>/<pod>.tar.gz`
Specific Example:
`.../__our_local_cocoapods/our-team/My-Cool-Package-5.9/3.3.746/My-Cool-Package-5.9.tar.gz`

> we need to understand how the list of packages is fetched

`artifactory-private` refers to a pod source we add manually as part of a setup script, whether running locally or on a CI machine:

```
bundle exec pod repo-art add artifactory-private https://artifactory.internal.domain.com/artifactory/api/pods/__our_local_cocoapods
```

If I understand the docs correctly, `pod repo-art add` is a [wrapper command](https://github.com/jfrog/cocoapods-art?tab=readme-ov-file#artifactory-configuration) supplied by the cocapods-art plugin that adds an Artifactory source to the Podfile. I believe at that point, when the Podfile is executed, the Artifactory url is added as a source and then used like a regular Podfile `source` declaration, which jives with what I'm seeing in the `Podfile.lock`.

**Reply by rarkins** [reply]
I've been struggling a bit to follow what we need to do.

It seems like there's a case where we need to do a git clone of the remote repo in order to read the files?

We also need to factor in security. If you're running self-hosted then you probably want to persist that clone indefinitely and just pull occasionally to refresh it. If it's in a multi-tenant environment like the Mend hosted app then we would need to delete that clone at the end of the repo run.

**Comment by mitchins** [comment]
I've been meaning to reply but it's been chaotic lately with releases going out.

@rarkins Yes, that is basically it.
If the source of cocoapods is not a CDN or Github v3 API, we'll need to use GIT.

I've followed the code from the crate data source - which leads me to believe it should be cleaned up by most build systems as it requests private cache dir, an excerpt below:
```
clonePath = upath.join(
        privateCacheDir(),
        PodDatasource.cacheDirFromUrl(url),
      );
```
Since Cocoapods is also being used it keeps its own data cache(`~/Library/Caches/CocoaPods`), the reasonable assumption is that the data belonging to the executing user for the build agent will be purged after the run?
Do you feel purging by renovate is required? From what I can see the crate datasource does not do that either?

@ElliottJW from what I can tell, based on the `source` line in your cocoapods, even though you are using artifactory for binary packages, the actual pod repository is pulled down via GIT (although artifactory can apparently support CDN too).

Outstanding questions:
1. How are we supposed to know the data-source url only understands GIT protocol and not a http CDN or Github API? (URLs can be tricky things as they mostly look the same, however we could infer what looks to be a GIT provider other than github, e.g. bamboo, azure, bitbucket). Injecting such a configuration for cocoapods seems like a huge architectural change outside of the scope of this ticket.
2. Is the private cache dir usage okay as-is?

**Reply by rarkins** [reply]
Actually there are two ways to achieve this:
1. Always allow git, and make sure to purge if it's multi-tenant
2. Make git opt-in (like cargo), then it doesn't need purging if so

**Reply by mitchins** [reply]
Thanks for that, do you know if:
1. Multi-tenant status
2. Allows GIT checkouts like cargo

are already exposed configuration flags?
Ideally not looking to change schemas or architecture.

In the absence of any existing or new flags, 1 has no blockers but the trick to ensure it's ALWAYS cleaned up even on failures.
This would look more like a final block and we'd need to make sure errors don't bail early.

Could you kindly explain by what mechanism the repository renovate checks out for the project itself is cleared with? Presumably the same concerns would exist in the project. Ideally the cleanup of this auxillary repo would be tied in with whatever renovate is doing as a whole.

**Reply by rarkins** [reply]
There isn't a generic "multi-tenant status" concept yet which is why there's point solutions like https://docs.renovatebot.com/self-hosted-configuration/#allowcustomcrateregistries

I suggest we do similar here as is done with crates, so that cleanup is not a big concern.

Renovate uses the fs package to remove clone directories.
